Export types documented for http_uri module

lucafavatella opened the pull request:
From the `http_uri` module documentation it may be intended that the listed types are exported by the module, and I have seen at least a couple of projects using those types as exported ([example](https://github.com/for-GET/jesse/blob/1.5.0/src/jesse.erl#L80)). I think it makes sense, hence this PR.

I consider this more a bug fix that a new feature, hence I based on `maint`. If you prefer I base on `master`, please let me know and I will be happy to do so.

-----

BTW I have not understood the `marker`s in the documentation e.g. https://github.com/erlang/otp/blob/OTP-20.2.4/lib/inets/doc/src/http_uri.xml#L78

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/erlang/otp?pullRequest=1724) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/erlang/otp?pullRequest=1724) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/erlang/otp?pullRequest=1724) it.</sub>

legoscia commented on lib/inets/src/http_lib/http_uri.erl:
```diff
@@ -61,16 +61,30 @@
+-type path() :: string() | unicode:unicode_binary().
+-type query() :: string() | unicode:unicode_binary().
+-type fragment() :: string() | unicode:unicode_binary().
+-type port_number() :: pos_integer().
```
This one could be defined as `inet:port_number()`.

peterdmv commented on lib/inets/doc/src/http_uri.xml:
```diff
@@ -63,8 +63,6 @@
        <item><p>Example: http, https</p></item>
        <tag><c>host() = string() | unicode:unicode_binary()</c></tag>
        <item><p></p></item>
-       <tag><c>port() = pos_integer()</c></tag>
```
I think the purpose of the author of this module was to describe the parts of the standard URI (scheme, userinfo, host, port, path, query, fragment). Thus, it would be better to keep 'port' with the type port_number().

lucafavatella replied:
Addressed in 6251070350e48c0e438cf848e2a44c5c5b73b7ae

peterdmv commented on lib/inets/src/http_lib/http_uri.erl:
```diff
@@ -61,16 +61,30 @@
+              query/0,
+              fragment/0]).
 
+-type uri() :: string() | unicode:unicode_binary().
```
Using the type unicode:unicode_binary() "implies" that the module can properly handle UTF-8 encoded binaries but unfortunately it can't. I would prefer using 'binary()' instead.

lucafavatella replied:
I understand this applies across the whole module. For some types - e.g. the ones that are output of functions - being proper UTF-8 binaries may apply but it looks too much effort ...

> As a side note, we are working on to remove application dependencies to the http_uri module and eventually obsolete it by the uri_string module in stdlib.

... especially considering the module is intended to be obsoleted.

Will propose change shortly.

lucafavatella replied:
Relevant past comment https://github.com/erlang/otp/pull/1322#issuecomment-299409428

lucafavatella replied:
Addressed in a31bdeb62a

lucafavatella commented:
> I would prefer these commits squashed into one commit.

I squashed the commits as one (the split history is still visible temporarily at https://github.com/lucafavatella/otp/tree/http_uri-types-pr-1724-bak ).

I added a new commit detailing http_uri:parse spec and doc, hence increasing slightly the scope of this PR. While addressing the comments I noticed the generation of atom for the scheme and I felt a warning is appropriate.

lucafavatella commented on lib/inets/src/http_lib/http_uri.erl:
```diff
@@ -81,13 +81,15 @@
 
+-type scheme_defaults() :: [{scheme(), default_scheme_port_number()}].
+-type scheme_validation_fun() :: fun((SchemeStr :: string() | binary()) ->
+                                            valid | {error, Reason :: term()}).
```
I see https://github.com/erlang/otp/blob/OTP-20.2.4/lib/inets/test/uri_SUITE.erl#L271-L274 says `non-fun scheme_validation_fun works as no option passed`. Though as a Developer I would expect incorrect value (non-fun) to cause some sort of error otherwise mistake in specification of value (e.g. mistake with some higher-order fun) could go unnoticed.

> So I propose to extend the type spec of scheme_validation_fun with the atom none.

I will follow as you propose in this PR.

lucafavatella replied:
Addressed in 9fc5b13f91

(BTW if you want to squash before merge, you may enable GitHub to do that for you/us as an alternative to the classic merge.)

peterdmv commented:
Thank you for your contribution!
